contracts/rewards/boosted-staking/BoostedDualVault.sol

```diff
@@ -0,0 +1,705 @@
+                    newRewardPerToken,
+                    false
+                );
+                uint256 platformEarned_ = _earned(
+                    _account,
+                    data.platformRewardPerTokenPaid,
+                    newPlatformRewardPerToken,
+                    true
+                );
+
+                // If earned == 0, then it must either be the initial stake, or an action in the
+                // same block, since new rewards unlock after each block.
+                if (earned_ > 0) {
+                    uint256 unlocked = earned_.mulTruncate(UNLOCK);
+                    uint256 locked = earned_ - unlocked;
+
+                    userRewards[_account].push(
+                        Reward({
+                            start: SafeCast.toUint64(LOCKUP + data.lastAction),
+                            finish: SafeCast.toUint64(LOCKUP + currentTime),
+                            rate: SafeCast.toUint128(locked / (currentTime - data.lastAction))
+                        })
+                    );
+
+                    userData[_account] = UserData({
+                        rewardPerTokenPaid: SafeCast.toUint128(newRewardPerToken),
+                        rewards: SafeCast.toUint128(unlocked + data.rewards),
+                        platformRewardPerTokenPaid: SafeCast.toUint128(newPlatformRewardPerToken),
+                        platformRewards: data.platformRewards + SafeCast.toUint128(platformEarned_),
+                        lastAction: currentTime64,
+                        rewardCount: data.rewardCount + 1
+                    });
+                } else {
+                    userData[_account] = UserData({
+                        rewardPerTokenPaid: SafeCast.toUint128(newRewardPerToken),
+                        rewards: data.rewards,
+                        platformRewardPerTokenPaid: SafeCast.toUint128(newPlatformRewardPerToken),
+                        platformRewards: data.platformRewards + SafeCast.toUint128(platformEarned_),
+                        lastAction: currentTime64,
+                        rewardCount: data.rewardCount
+                    });
+                }
+            }
+        } else if (_account != address(0)) {
+            // This should only be hit once, for first staker in initialisation case
+            userData[_account].lastAction = currentTime64;
+        }
+        _;
+    }
+
+    /** @dev Updates the boost for a given address, after the rest of the function has executed */
+    modifier updateBoost(address _account) {
+        _;
+        _setBoost(_account);
+    }
+
+    /***************************************
+                ACTIONS - EXTERNAL
+    ****************************************/
+
+    /**
+     * @dev Stakes a given amount of the StakingToken for the sender
+     * @param _amount Units of StakingToken
+     */
+    function stake(uint256 _amount)
+        external
+        override
+        updateReward(msg.sender)
```

Comment:
Why not have the `updateReward` and `updateBoost` modifiers on the internal `_stake` function passing the `beneficiary`? This is what the non-boosted `StakingRewards` contract does.

Reply:
Hmm, not sure. I think I just wanted to keep all external fns with the modifiers to avoid getting confused and potentially missing something. The updateBoost has to be called at the end of the external fn anyway, i think it may introduce a risk if there is logic in the external fn, to move it to internal

Reply:
Fair enough. I must admit I got confused reading the `StakingRewards` contract with the modified on the internal functions.